### ISSUE TITLE
Remove finalizer after marking resource as signed

### ISSUE THREAD

**Issue by lcarva** [tektoncd/chains#560]
Tekton Chains adds the finalizer `chains.tekton.dev` to TaskRun resources. My understanding is that the only reason for this is so Chains is "guaranteed" to have a chance to reconcile the resource before it gets deleted. All good here.

Once the resource is reconciled, the finalizer remains on the resource. When the resource is then deleted, the Chains controller processes the resource again. This is mostly a no-op because the controller ends early if the resource contains the annotation indicating it has been reconciled. There doesn't seem to be a good reason for the Chains controller to have to process the resource again once it has been properly reconciled. In fact, this cause unnecessary load on the controller, specially on clusters with a high number of TaskRuns.

Should we remove the finalizer from TaskRun resources once the resource has been marked as signed?

**Comment by wlynch** [comment]
Sounds good to me! Go for it. :)

**Comment by chuangw6** [comment]
Agree! Similarly, we should delete the finalizer from PipelineRun resources `chains.tekton.dev/pipelinerun` right?

I am wondering if @lcarva is working on this. If not, I am happy to try to open a PoC PR to fix this :D

**Comment by lcarva** [comment]
I am not currently working on this. If you have some time to do it, go for it! :pray:

**Comment by lcarva** [comment]
I don't think we can easily do this. First of all, there's no feature in `knative/pkg` about clearing the finalizer on-demand. This is done automatically by the framework after `FinalizeKind` is called: https://github.com/tektoncd/chains/blob/40b43e94287065ac6a8bc14a3ee6abfec888e7f6/vendor/github.com/tektoncd/pipeline/pkg/client/injection/reconciler/pipeline/v1beta1/taskrun/reconciler.go#L237-L244

Second, we can't call `clearFinalizer` because it's obviously not exported. Even if we were able to somehow reimplement that method, the finalizer would be re-added on the next resource update! https://github.com/tektoncd/chains/blob/40b43e94287065ac6a8bc14a3ee6abfec888e7f6/vendor/github.com/tektoncd/pipeline/pkg/client/injection/reconciler/pipeline/v1beta1/taskrun/reconciler.go#L226-L235

Given that there's no feature to support this in `knative/pkg` I don't think it's worth going down the path of hacking something to make it work.

I had a [discussion](https://knative.slack.com/archives/C93E33SN8/p1671738750746779) about this with someone in the knative slack (not sure how long the thread will be available for)

**Comment by lcarva** [comment]
Closing this as per my previous comment.
